lib/Dialect/TritonGPU/Transforms/WarpSpecialization/PartitionLoops.cpp

```diff
@@ -16,103 +19,331 @@ using namespace mlir;
-       llvm::make_early_inc_range(loop.getBody()->without_terminator())) {
-    if (!inPartition(&op)) {
-      op.dropAllUses();
-      op.erase();
-      continue;
+  auto isTensorResultFromOtherPartition = [&](int i) {
+    for (auto otherPartition : schedule.getPartitions()) {
+      if (otherPartition.getIndex() == partition->getIndex()) {
+        continue;
+      }
+      if (isTensorResultComputedBy(loop, i, &otherPartition, schedule)) {
+        return true;
+      }
     }
-    // Trace uses into the `scf.yield` to mark the corresponding iter arg as
-    // used.
-    for (OpOperand &use : op.getUses()) {
-      if (use.getOwner() == loop.getBody()->getTerminator())
-        toErase.reset(use.getOperandNumber());
+    return false;
+  };
+
+  SmallVector<LoopVarCategory> categories(loop.getNumRegionIterArgs());
+  for (auto [i, arg] : llvm::enumerate(loop.getRegionIterArgs())) {
+    if (llvm::any_of(arg.getUsers(), inPartition)) {
+      categories[i] = LoopVarCategory::Used;
+    } else if (isTensorResultFromOtherPartition(i)) {
+      categories[i] = LoopVarCategory::TensorResultFromOtherPartition;
+    } else {
+      categories[i] = LoopVarCategory::Unused;
     }
   }
-  for (auto [i, arg, result] :
-       llvm::enumerate(loop.getRegionIterArgs(), loop.getResults())) {
-    if (llvm::any_of(arg.getUsers(), inPartition) || !result.use_empty())
-      toErase.reset(i);
-    else if (toErase.test(i))
-      arg.dropAllUses();
+
+  return categories;
+}
+
+auto getLoopVarIndicesToKeep(
+    scf::ForOp loop, const Partition *partition,
+    const SmallVector<LoopVarCategory> loopVarCategories) {
+  SmallVector<size_t> indices;
+  SmallVector<int> reverseIndices(loop.getNumRegionIterArgs(), -1);
+  for (auto [i, arg] : llvm::enumerate(loop.getRegionIterArgs())) {
+    // For the default partition, keep non-tensor results used outside of the
+    // loop even if the corresponding loop variable is not used in that
+    // partition.
+    if (loopVarCategories[i] == LoopVarCategory::Used ||
+        (partition->getIndex() == 0 && !loop.getResult(i).use_empty() &&
+         loopVarCategories[i] !=
+             LoopVarCategory::TensorResultFromOtherPartition)) {
+      reverseIndices[i] = indices.size();
+      indices.push_back(i);
+    }
+  }
+  return std::make_pair(indices, reverseIndices);
+}
+
+auto getLoopVarIndicesToKeep(scf::ForOp loop, const Partition *partition,
+                             const WarpSchedule &schedule) {
+  auto loopVarCategories = classifyLoopVars(loop, partition, schedule);
+  return getLoopVarIndicesToKeep(loop, partition, loopVarCategories);
+}
+
+const Partition *getPartition(Operation *op, const WarpSchedule &schedule) {
+  while (op && !schedule.getPartition(op)) {
+    op = op->getParentOp();
+  }
+  if (op) {
+    return schedule.getPartition(op);
   }
-  eraseLoopCarriedValues(loop, std::move(toErase));
-  // Erase the schedule attributes.
-  WarpSchedule::eraseFrom(loop);
+  return nullptr;
```

Comment:
Should this assert and crash?

Reply:
It appears that we do hit this code path for some yield ops. One of them is an empty one `scf.yield"() : () -> ()` which I assume is added by the canonicalize pass. The other one is odd, if I try to dump such yield op, I get an error
```
Assertion failed: (mightHaveTerminator()), function getTerminator, file Block.cpp, line 247.
```